nexus/src/external_api/console_api.rs

```diff
@@ -91,12 +91,130 @@ pub async fn spoof_login(
+//
+// If the request has a signature, the query above will also contain SigAlg and
+// Signature params:
+//
+//   https://some.idp.test/auth/saml?SAMLRequest=...&RelayState=...&SigAlg=...&Signature=...
+//
+// SAMLRequest is base64 encoded zlib compressed XML, and RelayState can be
+// anything - Nexus currently encodes the referer header so that when SAML login
+// is successful the user can be sent back to where they were originally.
+//
+// The user will then authenticate with that IdP, and if successful will be
+// redirected back to the SP (Nexus) with a POST:
+//
+//   POST /login/{silo_name}/{provider_name}
+//
+// The body of this POST will contain a URL encoded payload that includes the
+// IdP's SAMLResponse plus optional relay state:
+//
+//   SAMLResponse=...&RelayState=...
+//
+// The RelayState will match what was sent as part of the initial redirect to
+// the IdP. The SAMLResponse will contain (among other things) the IdP's
+// assertion that this user has authenticated correctly, and provide information
+// about that user. Note that there is no Signature on the whole POST body, just
+// embedded in the SAMLResponse!
+//
+// The IdP's SAMLResponse will authenticate a subject, and from this external
+// subject a silo user has to be created or retrieved (depending on the Silo's
+// user provision type). After that, users will be redirected to the referer in
+// the relay state, or to `/organizations`.
+//
+// SAML logout flow
+// ----------------
+//
+// ** TODO SAML logout is currently unimplemented! **
+//
+// SAML logout is either SP initiated or IDP initiated.
+//
+// For SP inititated, a user will navigate to some yet-to-be-determined Nexus
+// URL. Something like:
+//
+//   GET /logout/{silo_name}/{provider_name}
+//
+// Nexus will redirect them to their IdP in order to perform the logout:
+//
+//   https://some.idp.test/auth/saml?SAMLRequest=...&RelayState=...
+//
+// where a LogoutRequest will be sent in the SAMLRequest parameter. If
+// successful, the IDP will redirect the user's browser back to:
+//
+//   POST /logout/{silo_name}/{provider_name}?SAMLRequest=...
+//
+// where there will be a LogoutRequest in the SAMLRequest parameter (where now
+// the IDP is requesting logout in the SP).
+//
+// For IDP inititated, the IDP can spontaneously POST a LogoutRequest to
+//
+//   /logout/{silo_name}/{provider_name}
+
 #[derive(Deserialize, JsonSchema)]
 pub struct LoginToProviderPathParam {
     pub silo_name: crate::db::model::Name,
     pub provider_name: crate::db::model::Name,
 }
 
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema)]
+pub struct RelayState {
+    pub referer: Option<String>,
+}
+
+impl RelayState {
+    pub fn to_encoded(&self) -> Result<String, anyhow::Error> {
+        Ok(base64::encode(serde_json::to_string(&self)?))
+    }
+
+    pub fn from_encoded(encoded: String) -> Result<Self, anyhow::Error> {
+        Ok(serde_json::from_str(&String::from_utf8(base64::decode(encoded)?)?)?)
```

Comment:
What do you think of using anyhow's `.context()` to put more information in this error?  I'm just worried if we somehow get an invalid value the error message might not say much about what went wrong.

Reply:
context in fe8c7566